### PR TITLE
PIMOB:2180 - Added CVV validation along with card scheme

### PR DESCRIPTION
**Pull request by chintan-soni-cko**
## Issue
[PIMOB-2180](https://checkout.atlassian.net/browse/PIMOB-2180)

## Proposed changes
- Functionality implementation: please read details ticket description
- Code level changes: Added cvv validator to validate the entered cvv value and expose the value of isEnteredCVVValid accordingly

**Code Refactor changes**
1. Rename package from cvvcomponent to cvvinputfield
2. Rename CVVComponentViewModel to CVVInputFieldViewModel
3. Rename CVVComponentViewModelFactory to CVVInputFieldViewModelFactory
Note: renaming and refactoring were done just for a more aligned naming convention for us which is input field

## Test Steps
1. Add a different scheme in CVVComponentConfig from MainActivity
2. Validate local logs for it whether isEnteredcvvValid value is updating properly according to card schemes

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

* [X] Reviewers assigned
* [X] I have performed a self-review of my code and manual testing
* [X] Lint and unit tests pass locally with my changes
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] I have added necessary documentation (if applicable)

## Further comments

_If this is a relatively large or complex change, kick off the discussion by explaining why you choose the solution you did and what alternatives you considered, etc..._

[PIMOB-2180]: https://checkout.atlassian.net/browse/PIMOB-2180?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: frames/src/test/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModelTest.kt [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    @Test
+    fun `when view model is initialised then initial state has empty cvv text`() {
+        // Then
+        assertTrue(viewModel.cvvInputFieldState.cvv.value.isEmpty())
+    }
+
+    @Test
+    fun `when view model is initialised then initial state has max cvv length updated as per card scheme from the cvvComponentConfig`() {
+        // Then
+        assertEquals(cvvComponentConfig.cardScheme.cvvLength.max(), viewModel.cvvInputFieldState.cvvLength.value)
```
Comment (jheng-hao-lin-cko) [inline frames/src/test/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModelTest.kt:98]: shall we add a test to cover that the `cvvLength` is actually taking effect?

**Review by jheng-hao-lin-cko — COMMENTED** [review]
Let's find a way to test different card schemes without changing the code.

File: frames/src/test/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModelTest.kt (hunk 1 above)
Comment (chintan-soni-cko) [reply frames/src/test/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModelTest.kt:98]: This test case covers cvvLength and is updated accordingly for the card scheme. However, it is a valid test to improve and it's part of the input field. We have not added any test cases for compose inputfield yet. We can add it to test improvements except if you think we can add validation for this in the viewmodel.

File: frames/src/main/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModel.kt [hunk 2]
```diff
@@ -1,34 +1,43 @@
...
     /**
      * Update mutable state of input field value.
      */
-    fun onCvvChange(text: String) {
-        if (text.isDigitsOnly()) {
-            cvvInputFieldState.text.value = text
+    fun onCvvChange(text: String) = with(text.isDigitsOnly()) {
+        cvvInputFieldState.cvv.value = text
+        validate()
+    }
```
[the remark this answers is not in the document]
Comment (jheng-hao-lin-cko) [reply frames/src/main/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModel.kt:33]: Gotcha that makes sense for the validation part.

There is another bit we could potentitally improve: we have `with(text.isDigitsOnly)` and get the boolean value from it, do we plan to anything with it? shall we guard if the text contains any non-digit char?

**Comment by jheng-hao-lin-cko** [comment]
Shall we support multiple card scheme on the same view for testing purpose?

**Comment by chintan-soni-cko** [comment]
> Let's find a way to test different card schemes without changing the code.

We will do it in sample app design

File: frames/src/main/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModel.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (chintan-soni-cko) [reply frames/src/main/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModel.kt:33]: In the styling part we have provided numeric keyboard and not allowing here non-digit characters using this `with(text.isDigitsOnly` function

**Comment by chintan-soni-cko** [comment]
> Shall we support multiple card scheme on the same view for testing purpose?



> Shall we support multiple card scheme on the same view for testing purpose?

Would it be fine if we pick this as part of [PIMOB-2176](https://checkout.atlassian.net/browse/PIMOB-2176) ticket and not blocking this PR? However, we can currently check from the sample app by changing the card scheme.

**Review by jheng-hao-lin-cko — APPROVED** [review]

File: frames/src/main/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModel.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (jheng-hao-lin-cko) [reply frames/src/main/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModel.kt:33]: so in this case we don't need `with(text.isDigitsOnly)` right?

[the remark this answers is not in the document]
Comment (chintan-soni-cko) [reply frames/src/main/java/com/checkout/frames/cvvinputfield/viewmodel/CVVInputFieldViewModel.kt:33]: Some keyboard or from the emulator still non-digits characters are visible when numeric keyboard is visible. So, it is nice to have to restrict it.

**Pull request merged**